src/device/socket/vsock.rs

```diff
@@ -101,24 +139,185 @@ impl<'a, H: Hal, T: Transport> VirtIOSocket<'a, H, T> {
+        }
+    }
+
+    /// Requests the credit and updates the credit in the current connection info.
+    pub fn request_credit(&mut self) -> Result {
+        let connection_info = self.connection_info()?;
+        let header = VirtioVsockHdr {
+            src_cid: self.guest_cid.into(),
+            dst_cid: connection_info.dst.cid.into(),
+            src_port: connection_info.src_port.into(),
+            dst_port: connection_info.dst.port.into(),
+            op: VirtioVsockOp::CreditRequest.into(),
+            ..Default::default()
+        };
+        self.send_packet_to_tx_queue(&header, &[])?;
+        self.wait_one_in_rx_queue();
+        let received_header = self.pop_header_only_packet_from_rx_queue()?;
+        match received_header.op()? {
+            VirtioVsockOp::CreditUpdate => {
+                if let Some(connection_info) = self.connection_info.as_mut() {
+                    connection_info.buf_alloc = received_header.buf_alloc.into();
+                    connection_info.fwd_cnt = received_header.fwd_cnt.into();
+                }
+                debug!("Connection info {:?}", self.connection_info);
+                Ok(())
+            }
+            VirtioVsockOp::Rst => Err(SocketError::ConnectionFailed.into()),
+            _ => Err(SocketError::InvalidOperation.into()),
+        }
+    }
+
+    /// Sends the buffer to the destination.
+    pub fn send(&mut self, buffer: &[u8]) -> Result {
+        self.request_credit()?;
+        let connection_info = self.connection_info()?;
+        let header = VirtioVsockHdr {
+            src_cid: self.guest_cid.into(),
+            dst_cid: connection_info.dst.cid.into(),
+            src_port: connection_info.src_port.into(),
+            dst_port: connection_info.dst.port.into(),
+            op: VirtioVsockOp::Rw.into(),
+            len: (buffer.len() as u32).into(),
+            buf_alloc: connection_info.buf_alloc.into(),
+            fwd_cnt: connection_info.fwd_cnt.into(),
+            ..Default::default()
+        };
+        self.send_packet_to_tx_queue(&header, buffer)
+    }
+
+    /// Receives the buffer from the destination.
+    /// Returns the actual size of the message.
+    pub fn recv(&mut self, buffer: &mut [u8]) -> Result<usize> {
+        let connection_info = self.connection_info()?;
+        let header = VirtioVsockHdr {
+            src_cid: self.guest_cid.into(),
+            dst_cid: connection_info.dst.cid.into(),
+            src_port: connection_info.src_port.into(),
+            dst_port: connection_info.dst.port.into(),
+            op: VirtioVsockOp::Rw.into(),
+            buf_alloc: connection_info.buf_alloc.into(),
+            fwd_cnt: connection_info.fwd_cnt.into(),
+            ..Default::default()
+        };
+        // TODO: We shouldn't need this send to trigger the fill of the rx queue.
+        self.send_packet_to_tx_queue(&header, &[])?;
+        self.wait_one_in_rx_queue();
+        let packet = self.pop_packet_from_rx_queue()?;
```

Comment:
How do you know that the next packet in the RX queue will be an RW, and not e.g. a CREDIT_UPDATE? The VirtIO spec says "It is also valid to send a VIRTIO_VSOCK_OP_CREDIT_UPDATE packet without previously receiving a VIRTIO_VSOCK_OP_CREDIT_REQUEST packet.", which suggests that we could receive one at any time.